drivers/ksz8851snl/ksz8851_netdev.c

```diff
@@ -0,0 +1,622 @@
+		uint16_t temp_irq = reg_get(dev, KSZ8851_PMECR_REG);
+		temp_irq |= (1 << 3);
+		temp_irq &= (~(0x000D << 2));
+		reg_set(dev, KSZ8851_PMECR_REG, temp_irq);
+	}
+
+	/* Link Change IRQ */
+	if (isr_mask & ISR_LCIS) {
+		/* Check and report what the current status of the link */
+		uint16_t link = reg_get(dev, KSZ8851_P1SR_REG);
+
+		netdev2_event_t event = (link & P1SR_LINK_GOOD) ?
+					NETDEV2_EVENT_LINK_UP :
+					NETDEV2_EVENT_LINK_DOWN;
+		if (link & P1SR_LINK_GOOD)
+			DEBUG(SECT_NAME "Link Status is Up\n");
+		else
+			DEBUG(SECT_NAME "Link Status is Down\n");
+
+		reg_set(dev, KSZ8851_ISR_REG, ISR_LCIS);           //TODO I need to look into unlocking here
+		dev->netdev.event_callback(netdev, event, NULL);        //could be wrong first variable
+	}
+	/* Packet Transmited IRQ */
+	if (isr_mask & ISR_TXIS) { /* TODO: Get rid of the Tx Irq */
+		reg_set(dev, KSZ8851_ISR_REG, ISR_TXIS);
+	}
+	/* Packet Received IRQ */
+	if (isr_mask & ISR_RXIS) {
+		/* Reset the IRQ Flag */
+		reg_set(dev, KSZ8851_ISR_REG, ISR_RXIS);
+		uint16_t frames_avail = (reg_get(dev, KSZ8851_RXFCTR_REG) >> 8);
+
+		while (frames_avail) {
+			unlock(dev);
+			netdev->event_callback(netdev, NETDEV2_EVENT_RX_COMPLETE, NULL);
+			lock(dev);
+			frames_avail--;
+		}
+	}
+	/* Restore IRQs and Release Mutex */
+	gpio_irq_enable(dev->int_pin);
+	reg_set(dev, KSZ8851_IER_REG, IER_DEFAULT_CONFIG);
+	unlock(dev);
+}
+
+static int ksz8851snl_reset(ksz8851snl_t *dev)
+{
+	uint16_t dev_id;
+
+	/* First attempt reset is a soft reset */
+	reg_set(dev, KSZ8851_GRR_REG, GRR_DEV_RESET);
+	xtimer_usleep(1000);
+	reg_set(dev, KSZ8851_GRR_REG, 0);
+
+	/* Check for a valid Chip */
+	dev_id = reg_get(dev, CHIP_ID_REG);
+
+    DEBUG(SECT_NAME"Chip id: 0x%04X\n", dev_id);
+
+	if ((dev_id & 0xFFF0) == CHIP_ID_8851_16)
+		return 0;
+
+	/* Second attempt reset is a hard reset */
+	gpio_clear(dev->rst);
+	xtimer_usleep(KSZ8851SNL_RST_DELAY);
+	gpio_set(dev->rst);
+	xtimer_usleep(KSZ8851SNL_BOOT_DELAY);
+
+	/* Check for a valid Chip */
+	dev_id = reg_get(dev, CHIP_ID_REG);
+
+	if ((dev_id & 0xFFF0) == CHIP_ID_8851_16)
+		return 0;
+
+	DEBUG(SECT_NAME "Reset Failed\n");
+	return 0;
+}
+
+static int ksz8851snl_init(netdev2_t *encdev)
+{
+	ksz8851snl_t *dev = (ksz8851snl_t*)encdev;
+	lock(dev);
+
+	/*** Setup IO ****/
+
+	/* Setup Chip Select */
+	gpio_init(dev->cs, GPIO_DIR_OUT, GPIO_PULLUP);
+	gpio_set(dev->cs);
+
+	/* Setup IRQ */
+	gpio_init(dev->int_pin, GPIO_DIR_IN, GPIO_PULLUP);
+	gpio_init_int(dev->int_pin, GPIO_PULLUP, GPIO_FALLING, ksz8851snl_pIsr, (void*)dev);
+
+	/* Setup RST */
+	gpio_init(dev->rst, GPIO_DIR_OUT, GPIO_PULLUP);
+	gpio_set(dev->rst);
+
+	/*** Setup SPI ***/
+	if (spi_init_master(dev->spi, SPI_CONF_FIRST_RISING, KSZ8851SNL_SPI_SPEED) < 0) return -1;
+
+	/* Soft reset and if that fails hard reset */
+	ksz8851snl_reset(dev);
+
+	/* Set Power Mode */
+	reg_set(dev, KSZ8851_PMECR_REG, PMECR_NORMAL);
+
+	/* Prototype Mac Address is 00:04:25:19:1A:3E */
```

Comment:
that is the other reason that I couldn't think of why i wanted to make the XPRO Ethernet its own folder because it is i2c to eeprom to get its individual MAC and that doesn't fit with the driver